### DISCUSSION TITLE
SWR and getserversideprops

### DISCUSSION THREAD

**Discussion by jesperordrup** [vercel/next.js#11116] [Help] [answered]
Goal: SSR page that also works as a SPA page that responds to a next/link. Is it good practice to use SWR with getsserversideprops (as it was with getinitialprops) or how?

**Comment by timneutkens** [comment]
> SSR page that also works as a SPA page that responds to a next/link

That's already the case, `next/link` client-side routing behavior is preserved when you use `getStaticProps` / `getServerSideProps`.

**Comment by jesperordrup** [comment]
> That's already the case, next/link client-side routing behavior is preserved when you use getStaticProps / getServerSideProps.

Thanks! 

As I understand it getServerSideProps code is always run on server even in client-side routing - correct? So is it a good idea to use SWR in the getServerSideProps or is it overkill?

**Reply by timneutkens** [reply]
You're referring to https://swr.now.sh/ right? Or `stale-while-revalidate` the header?

If you're referring to swr.now.sh, you can't use that in getServerSideProps because it's a React hook.

Next.js automatically fetches a JSON endpoint version of `getServerSideProps` in order to get the props needed for the page you're routing to.

**Reply by jesperordrup** [reply]
Thanks for clarifying, tim. I was sure that SWR was also avail serverside. My mistake.

**Reply by arihantverma** [reply]
@timneutkens I'm working on an application where four different routes have a same data requirement ( fetched for brevity of this argument, from `getInitialProps` ). Since you clarified above that JSON end point version of `getServerSideProps` or in `_app.js`'s case `getInitialProps` will always be triggered on route change, is there a way to NOT refetch it in those methods somehow ( since the data will already be present on the client from the previous route )? This might be outside of the purview of Next.js be just a web thing, but I'm not able to figure it out.

**Reply by gabrielsina** [reply]
@timneutkens regarding your message:
> https://github.com/vercel/next.js/discussions/11116#discussioncomment-1495

Does NEXT.js already support something to accomplish the use-case below?

--- 
You would like to intentionally expose all of your getServerSideProps functions to the client, so you could execute it from the client-side in case user is navigating between the same "page/route" - and have NEXT.js generate the new "pageProps" from the client-side. (Or at least manually configure which pages you would want to enable the above).

Example:
User is at "/search?k=bananas" and navigates to "/search?k=eggs".
User is at "/category/frozen-food" and navigates to "/category/fruits".

Benefits:
* Assume the async APIs you call from the getServerSideProps could be called directly from the client, hitting directly your API servers, instead of the node servers responsible for your NEXT.js client application.
* Assume you customize response by live user session and interaction with result items of the current page, so you don't really want to borrow cache strategies from next or even SWR.now.sh - hitting your API servers ASAP would always be ideal in such case.
* No need to manually refactor each page to use react-hook triggered fetching, and generically optimize navigation within the same route.


PS: Assume the developers do their diligence on "getServerSideProps" dependencies and sensitive data exposure on the pages for which you want to enable this optimization - at their own risk for bundle size side-effects and cybersecurity concerns.

Would love to hear your thoughts and learn more about nextjs current capabilities or future enhancements.

**Comment by jacobedawson** [comment, marked as answer]
SWR with Next.js / SSR example: https://github.com/vercel/swr#ssr-with-nextjs for googlers

**Reply by GarryOne** [reply]
Updated link: https://swr.vercel.app/docs/with-nextjs

**Comment by HamzaAmar** [comment]
Hello  @leerob @timneutkens  sometimes i want to use useSWR with getServerSideProps  and i don't like to go from getServerSideProps   to api route  this is the code that i do
```
export const getServerSideProps = async () => {
  try {
    await dbConnect()
    const products = await Product.find({})
    return { props: { products: JSON.parse(JSON.stringify(products)) } }
  } catch (err) {
    return { props: { error: JSON.parse(JSON.stringify(err)) } }
  }
}
```
But if i want to work with swr do i need to change the code and send it to api route because in useSWR i need the endpoint like `useSWR('/api/products" , fetcher ,{ initialData: props.products} )`  do i need to change the getServerSideProps  to somethings like that

```
export const getServerSideProps = async () => {
  try {
   const products = await fetch('api/products').then(res=>res.json())
   return { props: { products: JSON.parse(JSON.stringify(products)) } }
  } catch (err) {
    return { props: { error: JSON.parse(JSON.stringify(err)) } }
  }
}
```

**Reply by leerob** [reply]
That should be fine. You don't need to go to the API route, but make sure the data is formatted in the same way.

**Reply by HamzaAmar** [reply]
Thank you so much @leerob
